File: sdks/java/src/main/java/do_/mongo/Updates.java
package do_.mongo;

import java.util.Arrays;
import java.util.List;

/**
 * Utility class for creating MongoDB update operations.
 * <p>
 * This class provides static factory methods for all standard MongoDB update operators.
 * </p>
 *
 * <pre>{@code
 * import static do_.mongo.Updates.*;
 *
 * // Set a field
 * Bson update = set("status", "active");
 *
 * // Increment a field
 * Bson update = inc("count", 1);
 *
 * // Combine updates
 * Bson update = combine(
 *     set("status", "active"),
 *     inc("loginCount", 1),
 *     currentDate("lastLogin")
 * );
 * }</pre>
 */
public final class Updates {

    private Updates() {
        // Utility class
    }

    // ============================================================================
    // Field Update Operators
    // ============================================================================

    /**
     * Creates a $set update.
     *
     * @param fieldName the field name
     * @param value     the new value
     * @return the update
     */
    public static Bson set(String fieldName, Object value) {
        return () -> new Document("$set", new Document(fieldName, value));
    }

    /**
     * Creates a $setOnInsert update.
     *
     * @param fieldName the field name
     * @param value     the value to set on insert
     * @return the update
     */
    public static Bson setOnInsert(String fieldName, Object value) {
        return () -> new Document("$setOnInsert", new Document(fieldName, value));
    }

    /**
     * Creates an $unset update.
     *
     * @param fieldName the field to unset
     * @return the update
     */
    public static Bson unset(String fieldName) {
        return () -> new Document("$unset", new Document(fieldName, ""));
    }

    /**
     * Creates a $rename update.
     *
     * @param fieldName    the current field name
     * @param newFieldName the new field name
     * @return the update
     */
    public static Bson rename(String fieldName, String newFieldName) {
        return () -> new Document("$rename", new Document(fieldName, newFieldName));
    }

    // ============================================================================
    // Numeric Update Operators
    // ============================================================================

    /**
     * Creates an $inc update.
     *
     * @param fieldName the field name
     * @param number    the amount to increment
     * @return the update
     */
    public static Bson inc(String fieldName, Number number) {
        return () -> new Document("$inc", new Document(fieldName, number));
    }

    /**
     * Creates a $mul update.
     *
     * @param fieldName the field name
     * @param number    the multiplier
     * @return the update
     */
    public static Bson mul(String fieldName, Number number) {
        return () -> new Document("$mul", new Document(fieldName, number));
    }

    /**
     * Creates a $min update.
     *
     * @param fieldName the field name
     * @param value     the minimum value
     * @return the update
     */
    public static Bson min(String fieldName, Object value) {
        return () -> new Document("$min", new Document(fieldName, value));
    }

    /**
     * Creates a $max update.
     *
     * @param fieldName the field name
     * @param value     the maximum value
     * @return the update
     */
    public static Bson max(String fieldName, Object value) {
        return () -> new Document("$max", new Document(fieldName, value));
    }

    // ============================================================================
    // Date Update Operators
    // ============================================================================

    /**
     * Creates a $currentDate update that sets the field to the current date.
     *
     * @param fieldName the field name
     * @return the update
     */
    public static Bson currentDate(String fieldName) {
        return () -> new Document("$currentDate", new Document(fieldName, true));
    }

    /**
     * Creates a $currentDate update that sets the field to the current timestamp.
     *
     * @param fieldName the field name
     * @return the update
     */
    public static Bson currentTimestamp(String fieldName) {
        return () -> new Document("$currentDate",
                new Document(fieldName, new Document("$type", "timestamp")));
    }

    // ============================================================================
    // Array Update Operators
    // ============================================================================

    /**
     * Creates a $push update.
     *
     * @param fieldName the field name
     * @param value     the value to push
     * @return the update
     */
    public static Bson push(String fieldName, Object value) {
        return () -> new Document("$push", new Document(fieldName, value));
    }

    /**
     * Creates a $push update with $each.
     *
     * @param fieldName the field name
     * @param values    the values to push
     * @return the update
     */
    public static Bson pushEach(String fieldName, Object... values) {
        return pushEach(fieldName, Arrays.asList(values));
    }

    /**
     * Creates a $push update with $each.
     *
     * @param fieldName the field name
     * @param values    the values to push
     * @return the update
     */
    public static Bson pushEach(String fieldName, List<?> values) {
        return () -> new Document("$push",
                new Document(fieldName, new Document("$each", values)));
    }

    /**
     * Creates a $push update with $each and $slice.
     *
     * @param fieldName the field name
     * @param values    the values to push
     * @param slice     the slice limit
     * @return the update
     */
    public static Bson pushEach(String fieldName, List<?> values, int slice) {
        return () -> new Document("$push",
                new Document(fieldName, new Document("$each", values).append("$slice", slice)));
    }

    /**
     * Creates a $push update with $each, $slice, and $sort.
     *
     * @param fieldName     the field name
     * @param values        the values to push
     * @param slice         the slice limit
     * @param sortField     the field to sort by
     * @param sortDirection 1 for ascending, -1 for descending
     * @return the update
     */
    public static Bson pushEach(String fieldName, List<?> values, int slice, String sortField, int sortDirection) {
        return () -> new Document("$push",
                new Document(fieldName, new Document("$each", values)
                        .append("$slice", slice)
                        .append("$sort", new Document(sortField, sortDirection))));
    }

    /**
     * Creates an $addToSet update.
     *
     * @param fieldName the field name
     * @param value     the value to add
     * @return the update
     */
    public static Bson addToSet(String fieldName, Object value) {
        return () -> new Document("$addToSet", new Document(fieldName, value));
    }

    /**
     * Creates an $addToSet update with $each.
     *
     * @param fieldName the field name
     * @param values    the values to add
     * @return the update
     */
    public static Bson addEachToSet(String fieldName, Object... values) {
        return addEachToSet(fieldName, Arrays.asList(values));
    }

    /**
     * Creates an $addToSet update with $each.
     *
     * @param fieldName the field name
     * @param values    the values to add
     * @return the update
     */
    public static Bson addEachToSet(String fieldName, List<?> values) {
        return () -> new Document("$addToSet",
                new Document(fieldName, new Document("$each", values)));
    }

    /**
     * Creates a $pop update to remove the first element.
     *
     * @param fieldName the field name
     * @return the update
     */
    public static Bson popFirst(String fieldName) {
        return () -> new Document("$pop", new Document(fieldName, -1));
    }

    /**
     * Creates a $pop update to remove the last element.
     *
     * @param fieldName the field name
     * @return the update
     */
    public static Bson popLast(String fieldName) {
        return () -> new Document("$pop", new Document(fieldName, 1));
    }

    /**
     * Creates a $pull update.
     *
     * @param fieldName the field name
     * @param value     the value to remove
     * @return the update
     */
    public static Bson pull(String fieldName, Object value) {
        return () -> new Document("$pull", new Document(fieldName, value));
    }

    /**
     * Creates a $pull update with a filter.
     *
     * @param fieldName the field name
     * @param filter    the filter for elements to remove
     * @return the update
     */
    public static Bson pullByFilter(String fieldName, Bson filter) {
        return () -> new Document("$pull", new Document(fieldName, filter.toBsonDocument()));
    }

    /**
     * Creates a $pullAll update.
     *
     * @param fieldName the field name
     * @param values    the values to remove
     * @return the update
     */
    public static Bson pullAll(String fieldName, Object... values) {
        return pullAll(fieldName, Arrays.asList(values));
    }

    /**
     * Creates a $pullAll update.
     *
     * @param fieldName the field name
     * @param values    the values to remove
     * @return the update
     */
    public static Bson pullAll(String fieldName, List<?> values) {
        return () -> new Document("$pullAll", new Document(fieldName, values));
    }

    // ============================================================================
    // Bitwise Update Operators
    // ============================================================================

    /**
     * Creates a $bit update with AND.
     *
     * @param fieldName the field name
     * @param value     the bitmask
     * @return the update
     */
    public static Bson bitwiseAnd(String fieldName, long value) {
        return () -> new Document("$bit", new Document(fieldName, new Document("and", value)));
    }

    /**
     * Creates a $bit update with OR.
     *
     * @param fieldName the field name
     * @param value     the bitmask
     * @return the update
     */
    public static Bson bitwiseOr(String fieldName, long value) {
        return () -> new Document("$bit", new Document(fieldName, new Document("or", value)));
    }

    /**
     * Creates a $bit update with XOR.
     *
     * @param fieldName the field name
     * @param value     the bitmask
     * @return the update
     */
    public static Bson bitwiseXor(String fieldName, long value) {
        return () -> new Document("$bit", new Document(fieldName, new Document("xor", value)));
    }

    // ============================================================================
    // Combination
    // ============================================================================

    /**
     * Combines multiple updates into a single update.
     *
     * @param updates the updates to combine
     * @return the combined update
     */
    public static Bson combine(Bson... updates) {
        return combine(Arrays.asList(updates));
    }

    /**
     * Combines multiple updates into a single update.
     *
     * @param updates the updates to combine
     * @return the combined update
     */
    public static Bson combine(List<Bson> updates) {
        return () -> {
            Document combined = new Document();
            for (Bson update : updates) {
                Document doc = update.toBsonDocument();
                for (String key : doc.keySet()) {
                    if (combined.containsKey(key)) {
                        // Merge operators with the same key
                        Object existing = combined.get(key);
                        Object newValue = doc.get(key);
                        if (existing instanceof Document && newValue instanceof Document) {
                            ((Document) existing).putAll((Document) newValue);
                        } else {
                            combined.put(key, newValue);
                        }
                    } else {
                        combined.put(key, doc.get(key));
                    }
                }
            }
            return combined;
        };
    }
}

File: sdks/kotlin/src/main/kotlin/do_/mongo/Updates.kt
package do_.mongo

/**
 * Update builders for MongoDB operations.
 *
 * Example usage:
 * ```kotlin
 * import do_.mongo.Updates.*
 *
 * // Simple set
 * val update = set("name", "Jane")
 *
 * // Multiple updates
 * val updates = combine(
 *     set("name", "Jane"),
 *     inc("age", 1),
 *     unset("oldField")
 * )
 *
 * // Using DSL
 * val update = update {
 *     "name" setTo "Jane"
 *     "age" inc 1
 *     "lastUpdated" currentDate true
 * }
 * ```
 */
object Updates {

    // =========================================================================
    // Field Update Operators
    // =========================================================================

    /**
     * Creates a $set update.
     */
    fun set(field: String, value: Any?): Bson =
        Document("\$set", Document(field, value))

    /**
     * Creates a $set update for multiple fields.
     */
    fun set(vararg pairs: Pair<String, Any?>): Bson =
        Document("\$set", Document(pairs.toMap()))

    /**
     * Creates a $setOnInsert update.
     */
    fun setOnInsert(field: String, value: Any?): Bson =
        Document("\$setOnInsert", Document(field, value))

    /**
     * Creates an $unset update.
     */
    fun unset(vararg fields: String): Bson =
        Document("\$unset", Document(fields.associateWith { "" }))

    /**
     * Creates a $rename update.
     */
    fun rename(field: String, newName: String): Bson =
        Document("\$rename", Document(field, newName))

    /**
     * Creates an $inc update.
     */
    fun inc(field: String, amount: Number): Bson =
        Document("\$inc", Document(field, amount))

    /**
     * Creates a $mul update.
     */
    fun mul(field: String, multiplier: Number): Bson =
        Document("\$mul", Document(field, multiplier))

    /**
     * Creates a $min update.
     */
    fun min(field: String, value: Any?): Bson =
        Document("\$min", Document(field, value))

    /**
     * Creates a $max update.
     */
    fun max(field: String, value: Any?): Bson =
        Document("\$max", Document(field, value))

    /**
     * Creates a $currentDate update.
     */
    fun currentDate(field: String): Bson =
        Document("\$currentDate", Document(field, true))

    /**
     * Creates a $currentDate update with timestamp type.
     */
    fun currentTimestamp(field: String): Bson =
        Document("\$currentDate", Document(field, Document("\$type", "timestamp")))

    // =========================================================================
    // Array Update Operators
    // =========================================================================

    /**
     * Creates a $push update.
     */
    fun push(field: String, value: Any?): Bson =
        Document("\$push", Document(field, value))

    /**
     * Creates a $push update with modifiers.
     */
    fun pushEach(field: String, values: List<Any?>, position: Int? = null, slice: Int? = null, sort: Any? = null): Bson {
        val eachDoc = Document("\$each", values)
        position?.let { eachDoc.append("\$position", it) }
        slice?.let { eachDoc.append("\$slice", it) }
        sort?.let { eachDoc.append("\$sort", it) }
        return Document("\$push", Document(field, eachDoc))
    }

    /**
     * Creates an $addToSet update.
     */
    fun addToSet(field: String, value: Any?): Bson =
        Document("\$addToSet", Document(field, value))

    /**
     * Creates an $addToSet update with multiple values.
     */
    fun addEachToSet(field: String, values: List<Any?>): Bson =
        Document("\$addToSet", Document(field, Document("\$each", values)))

    /**
     * Creates a $pop update (remove first element).
     */
    fun popFirst(field: String): Bson =
        Document("\$pop", Document(field, -1))

    /**
     * Creates a $pop update (remove last element).
     */
    fun popLast(field: String): Bson =
        Document("\$pop", Document(field, 1))

    /**
     * Creates a $pull update.
     */
    fun pull(field: String, value: Any?): Bson =
        Document("\$pull", Document(field, value))

    /**
     * Creates a $pull update with a filter.
     */
    fun pull(field: String, filter: Bson): Bson =
        Document("\$pull", Document(field, filter.toBsonDocument()))

    /**
     * Creates a $pullAll update.
     */
    fun pullAll(field: String, values: List<Any?>): Bson =
        Document("\$pullAll", Document(field, values))

    // =========================================================================
    // Bitwise Update Operators
    // =========================================================================

    /**
     * Creates a bitwise AND update.
     */
    fun bitwiseAnd(field: String, value: Long): Bson =
        Document("\$bit", Document(field, Document("and", value)))

    /**
     * Creates a bitwise OR update.
     */
    fun bitwiseOr(field: String, value: Long): Bson =
        Document("\$bit", Document(field, Document("or", value)))

    /**
     * Creates a bitwise XOR update.
     */
    fun bitwiseXor(field: String, value: Long): Bson =
        Document("\$bit", Document(field, Document("xor", value)))

    // =========================================================================
    // Combination
    // =========================================================================

    /**
     * Combines multiple updates into one.
     */
    fun combine(vararg updates: Bson): Bson {
        val combined = Document()
        for (update in updates) {
            val doc = update.toBsonDocument()
            for ((key, value) in doc) {
                val existing = combined[key]
                if (existing is Document && value is Document) {
                    existing.putAll(value)
                } else {
                    combined[key] = value
                }
            }
        }
        return combined
    }

    /**
     * Combines a collection of updates into one.
     */
    fun combine(updates: Collection<Bson>): Bson = combine(*updates.toTypedArray())
}

// =========================================================================
// DSL Support
// =========================================================================

/**
 * Update DSL builder.
 */
@DocumentDsl
class UpdateBuilder {
    private val updates = mutableListOf<Bson>()

    infix fun String.setTo(value: Any?) {
        updates.add(Updates.set(this, value))
    }

    infix fun String.inc(amount: Number) {
        updates.add(Updates.inc(this, amount))
    }

    infix fun String.mul(multiplier: Number) {
        updates.add(Updates.mul(this, multiplier))
    }

    infix fun String.min(value: Any?) {
        updates.add(Updates.min(this, value))
    }

    infix fun String.max(value: Any?) {
        updates.add(Updates.max(this, value))
    }

    infix fun String.push(value: Any?) {
        updates.add(Updates.push(this, value))
    }

    infix fun String.addToSet(value: Any?) {
        updates.add(Updates.addToSet(this, value))
    }

    infix fun String.pull(value: Any?) {
        updates.add(Updates.pull(this, value))
    }

    fun String.unset() {
        updates.add(Updates.unset(this))
    }

    infix fun String.currentDate(set: Boolean) {
        if (set) updates.add(Updates.currentDate(this))
    }

    infix fun String.renameTo(newName: String) {
        updates.add(Updates.rename(this, newName))
    }

    internal fun toBson(): Bson = Updates.combine(updates)
}

/**
 * Creates an update using DSL syntax.
 */
inline fun update(block: UpdateBuilder.() -> Unit): Bson =
    UpdateBuilder().apply(block).toBson()
